### ISSUE TITLE
CA2000: False-positive when IDisposable is passed to wrapper object

### ISSUE THREAD

**Issue by skrysmanski** [dotnet/roslyn-analyzers#4636]
### Analyzer

**Diagnostic ID**: [CA2000](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2000): `Dispose objects before losing scope`

### Analyzer source

**NuGet Package**: [Microsoft.CodeAnalysis.NetAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.NetAnalyzers)

**Version**: 5.0.1 (Latest)

### Describe the bug

CA2000 is reported (for `new HttpClientHandler()`) in the following code:

```c#
private static HttpClient CreateHttpClient()
{
    HttpClientHandler? handler = null;
    try
    {
        handler = new HttpClientHandler();
        handler.CheckCertificateRevocationList = true;

        return new HttpClient(handler);
    }
    catch (Exception)
    {
        handler?.Dispose();
        throw;
    }
}
```

Error message:

> error CA2000: Call System.IDisposable.Dispose on object created by 'new HttpClientHandler()' before all references to it are out of scope

Maybe I'm missing something here but I can't figure out a way how to fix this (in my "opinion" there is none - so it's a false-positive).

### Steps To Reproduce

1. Create .NET 5 console project with `<Nullable>enable</Nullable>` and `<AnalysisMode>AllEnabledByDefault</AnalysisMode>`
1. Add the code above to the project
1. Compile

### Expected behavior

No CA2000 should be reported.

### Actual behavior

CA2000 is reported.

## Additional context

Both `HttpClient` as well as `HttpClientHandler` implement `IDisposable` (if that matters).

I've noticed that the rule description states that the warning can be suppressed, if:

> The method that raised the warning returns an IDisposable object that wraps your object

This is the case in my example. However, I've also noticed that this case is automatically suppressed for certain types listed under [Special Cases](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2000#special-cases).

Maybe `HttpClientHandler` should be added to this list.

Or maybe CA2000 should never be raised (perhaps configurable) in this situation where the IDisposable is passed to a wrapper object (either as constructor parameter or as property) and this wrapper object is returned.

**Comment by huoyaoyuan** [comment]
CA2000 is very problematic because there's no correct annotation for ownership and wrapping. The inferred annotation is probably wrong for service consumers.

**Comment by mavasani** [comment]
Mark as a duplicate of ownership transfer isue.
